### DISCUSSION TITLE
[Basic Tutorial 2] How to Package a Kernel using AnyKernel 3

### DISCUSSION THREAD

**Discussion by Coconutat** [tiann/KernelSU#954] [Show and tell]
# How to Package a Kernel using AnyKernel 3
As the title suggests, this is still a basic tutorial. Experienced friends can skip ahead.

***

## 1. What is AnyKernel 3?
[AnyKernel3](https://github.com/osm0sis/AnyKernel3) is a flashing template for Android device kernels developed by [osm0sis](https://github.com/osm0sis), a developer on the XDA forum. In fact, it is very powerful and can accommodate many custom operations.

For example:
1. Modify only the kernel portion without interfering with others (e.g., Magisk can be preserved).
2. Customize kernel installation (e.g., let users choose the actions during installation).
3. Automatically patch vbmeta (depending on whether you enable this feature), and more.

This tutorial focuses only on the basics and does not delve into more complex content. So, don't worry; there's nothing too difficult.

The prerequisite for using AnyKernel 3 is that your device has a custom RECOVERY (e.g., TWRP, OrangeFox, or supports adb sideload). If you do not meet this condition, you can try unpacking/packaging boot.img for kernel replacement. If you have SU privileges, you can use a kernel flashing app and then use AnyKernel 3 for flashing (e.g., Kernel Flasher, Franco Kernel Manager, EX Kernel Manager).

***

## 2. Using AnyKernel 3
On your computer, use the following command to clone this project:
```
git clone https://github.com/osm0sis/AnyKernel3.git
```
Then, navigate to the root directory of AnyKernel 3.

Next, let's understand the basic directory structure of AnyKernel 3:
```
META-INF
modules
patch
ramdisk
tools
LICENSE
README.md
anykernel.sh
```
As a beginner, you don't need to worry about the other files for now; just focus on **anykernel.sh** and **LICENSE**.

Open **anykernel.sh** with your favorite text editor. You'll see a lot of content, but let me simplify it for you. Here's the streamlined version:

```
### AnyKernel3 Ramdisk Mod Script
## osm0sis @ xda-developers

### AnyKernel setup
# global properties
properties() { '
kernel.string=ExampleKernel by osm0sis @ xda-developers
do.devicecheck=1
do.modules=0
do.systemless=1
do.cleanup=1
do.cleanuponabort=0
device.name1=
device.name2=
device.name3=
device.name4=
device.name5=
supported.versions=
supported.patchlevels=
supported.vendorpatchlevels=
'; } # end properties

### AnyKernel install
## boot files attributes
boot_attributes() {
set_perm_recursive 0 0 755 644 $ramdisk/*;
set_perm_recursive 0 0 750 750 $ramdisk/init* $ramdisk/sbin;
} # end attributes

# boot shell variables
block= XXXX ;
is_slot_device=0;
ramdisk_compression=auto;
patch_vbmeta_flag=auto;

# import functions/variables and setup patching - see for reference (DO NOT REMOVE)
. tools/ak3-core.sh;

# boot install
dump_boot; # use split_boot to skip ramdisk unpack, e.g., for devices with init_boot ramdisk

write_boot; # use flash_boot to skip ramdisk repack, e.g., for devices with init_boot ramdisk
## end boot install
```

Here's what you need to know from this script:
1. `kernel.string=`: This is where you should enter the name of your kernel. You can write anything you like.
2. `do.devicecheck=`: Enter 1 to enable device checking or 0 to disable it. This is used to check the device name, which corresponds to `device.name1` to `device.name5` below. If the device name doesn't match, the kernel won't flash.
3. `device.name1` to `device.name5`: Here, you can enter your device's name. Note that some phones have development codenames, while others have specific device names. Use these parameters with `do.devicecheck` enabled.
4. `block= XXXX`: This should specify the partition where the kernel should be flashed, typically "boot."
5. `is_slot_device=`: Enter 1 to enable, 0 to disable, or "auto" for automatic detection. Use this for devices with AB slots (Virtual A/B or A/B partitioning).
6. `patch_vbmeta_flag=`: Enter 1 to enable, 0 to disable, or "auto" for automatic detection. This determines whether to patch vbmeta to disable AVB verification. This depends on your device, so you need to make the determination yourself.

With this information, you can modify the provided template script to create your own kernel flash package. Here's a generic script you can use:

```
### AnyKernel3 Ramdisk Mod Script
## osm0sis @ xda-developers

### AnyKernel setup
# global properties
properties() { '
kernel.string=ExampleKernel by osm0sis @ xda-developers # Enter your kernel name here
do.devicecheck=0 # Disable device checking
do.modules=0
do.systemless=1
do.cleanup=1
do.cleanuponabort=0
supported.versions=
supported.patchlevels=
supported.vendorpatchlevels=
'; } # end properties

### AnyKernel install
## boot files attributes
boot_attributes() {
set_perm_recursive 0 0 755 644 $ramdisk/*;
set_perm_recursive 0 0 750 750 $ramdisk/init* $ramdisk/sbin;
} # end attributes

# boot shell variables
block=boot; # Partition to flash is "boot"
is_slot_device=auto; # Automatically detect AB slots
ramdisk_compression=auto;
patch_vbmeta_flag=auto;

# import functions/variables and setup patching - see for reference (DO NOT REMOVE)
. tools/ak3-core.sh;

# boot install
dump_boot; # Use split_boot to skip ramdisk unpack, e.g., for devices with init_boot ramdisk

write_boot; # Use flash_boot to skip ramdisk repack, e.g., for devices with init_boot ramdisk
## end boot install
```

**Note that this script is just a template. You still need to modify and verify its contents based on the instructions above!**

***

## 3. Packaging Your Kernel
Copy your kernel into the root directory of AnyKernel 3. Assuming your kernel's name is **Image.gz**, your AnyKernel 3 root directory should look like this:

```
META-INF
modules
patch
ramdisk
tools
LICENSE
README.md
anykernel.sh
Image.gz
```

Then, compress all the contents in the root directory into a zip file. Please remember to use **zip**, not rar, 7z, tar, or any other format!!!

Now, you can flash and test your kernel using TWRP.

Regarding the **LICENSE**:
The LICENSE file is a license agreement, which means AnyKernel 3 is legally protected, and you must comply with the terms of the license. Please strictly adhere to the restrictions outlined in the LICENSE.

***

## 4. Conclusion
This is just a basic tutorial on using AnyKernel 3. If you have any better ideas or suggestions, please feel free to share them in the comments.

## 5. Thanks
@osm0sis
> Author's Page: [osm0sis](https://github.com/osm0sis)

**Comment by anuragpindoriya** [comment]
@Coconutat i suggest  make also tutorial on unpacking/packaging boot.img for kernel

**Comment by fanmingyi** [comment]
this thread is extememly well-writed

**Comment by axshhayy** [comment]
Well explained 🙌🏻
